File: android/app/src/test/java/miko/todayinspacehistory/services/ImageProviderServiceTests.kt
package miko.todayinspacehistory.services

import kotlinx.coroutines.test.runTest
import miko.todayinspacehistory.data.Errors
import miko.todayinspacehistory.data.ImageProviderServiceImpl
import miko.todayinspacehistory.support.MockNetworkClient
import miko.todayinspacehistory.support.TestFixtures
import org.junit.Assert.assertEquals
import org.junit.Assert.assertTrue
import org.junit.Assert.fail
import org.junit.Test

class ImageProviderServiceTests {
    @Test
    fun prefersAnniversaryItems() = runTest {
        val client = MockNetworkClient()
        val anniversary = TestFixtures.item(
            title = "Anniversary",
            dateCreated = TestFixtures.ANNIVERSARY_DATE_CREATED,
            href = "https://example.com/a.json",
        )
        val other = TestFixtures.item(
            title = "Other",
            dateCreated = TestFixtures.OTHER_DATE_CREATED,
            href = "https://example.com/o.json",
        )
        client.searchResult = Result.success(TestFixtures.apiResponse(listOf(other, anniversary)))
        client.fetchImagesResult = Result.success(listOf("https://images.example.com/large.jpg"))

        val service = ImageProviderServiceImpl(client = client, now = { TestFixtures.july20_2026() })
        val result = service.loadTodaysImage()

        assertEquals("Anniversary", result.item.data?.first()?.title)
        assertEquals(listOf("https://example.com/a.json"), client.fetchImagesUrls)
        assertEquals(listOf("https://images.example.com/large.jpg"), result.imageUrls)
    }

    @Test
    fun fallsBackWhenNoAnniversaryMatches() = runTest {
        val client = MockNetworkClient()
        val other = TestFixtures.item(
            title = "Fallback",
            dateCreated = TestFixtures.OTHER_DATE_CREATED,
            href = "https://example.com/f.json",
        )
        client.searchResult = Result.success(TestFixtures.apiResponse(listOf(other)))
        client.fetchImagesResult = Result.success(listOf("http://images.example.com/medium.jpg"))

        val service = ImageProviderServiceImpl(client = client, now = { TestFixtures.july20_2026() })
        val result = service.loadTodaysImage()

        assertEquals("Fallback", result.item.data?.first()?.title)
        assertEquals(listOf("https://images.example.com/medium.jpg"), result.imageUrls)
    }

    @Test
    fun throwsNoItems() = runTest {
        val client = MockNetworkClient()
        client.searchResult = Result.success(TestFixtures.apiResponse(emptyList()))
        val service = ImageProviderServiceImpl(client = client, now = { TestFixtures.july20_2026() })

        try {
            service.loadTodaysImage()
            fail("Expected noItems")
        } catch (error: Errors.ImageProvider.NoItems) {
            assertTrue(true)
        }
    }

    @Test
    fun throwsMissingAssetURL() = runTest {
        val client = MockNetworkClient()
        val item = TestFixtures.item(
            dateCreated = TestFixtures.ANNIVERSARY_DATE_CREATED,
            href = null,
        )
        client.searchResult = Result.success(TestFixtures.apiResponse(listOf(item)))
        val service = ImageProviderServiceImpl(client = client, now = { TestFixtures.july20_2026() })

        try {
            service.loadTodaysImage()
            fail("Expected missingAssetURL")
        } catch (error: Errors.ImageProvider.MissingAssetURL) {
            assertTrue(true)
        }
    }

    @Test
    fun rewritesOnlyHTTPScheme() = runTest {
        val client = MockNetworkClient()
        val item = TestFixtures.item(
            dateCreated = TestFixtures.ANNIVERSARY_DATE_CREATED,
            href = "https://example.com/a.json",
        )
        client.searchResult = Result.success(TestFixtures.apiResponse(listOf(item)))
        client.fetchImagesResult = Result.success(
            listOf(
                "http://cdn.example.com/large.jpg",
                "https://cdn.example.com/original.jpg",
            ),
        )

        val service = ImageProviderServiceImpl(client = client, now = { TestFixtures.july20_2026() })
        val result = service.loadTodaysImage()

        assertEquals(
            listOf(
                "https://cdn.example.com/large.jpg",
                "https://cdn.example.com/original.jpg",
            ),
            result.imageUrls,
        )
    }
}

File: ios/TodayInSpaceHistoryTests/Services/ImageProviderServiceTests.swift
//
//  ImageProviderServiceTests.swift
//  TodayInSpaceHistoryTests
//

import XCTest
@testable import TodayInSpaceHistory

final class ImageProviderServiceTests: XCTestCase {
    func testPrefersAnniversaryItems() async throws {
        let client = MockNetworkClient()
        let anniversary = TestFixtures.item(
            title: "Anniversary",
            dateCreated: TestFixtures.todaysAnniversaryDateCreated,
            href: "https://example.com/a.json"
        )
        let other = TestFixtures.item(
            title: "Other",
            dateCreated: TestFixtures.nonAnniversaryDateCreated,
            href: "https://example.com/o.json"
        )
        client.searchResult = .success(TestFixtures.apiResponse(items: [other, anniversary]))
        client.fetchImagesResult = .success(["https://images.example.com/large.jpg"])

        let service = ImageProviderService(client: client)
        let result = try await service.loadTodaysImage()

        XCTAssertEqual(result.item.data?.first?.title, "Anniversary")
        XCTAssertEqual(client.fetchImagesURLs, ["https://example.com/a.json"])
        XCTAssertEqual(result.imageURLs, ["https://images.example.com/large.jpg"])
    }

    func testFallsBackWhenNoAnniversaryMatches() async throws {
        let client = MockNetworkClient()
        let other = TestFixtures.item(
            title: "Fallback",
            dateCreated: TestFixtures.nonAnniversaryDateCreated,
            href: "https://example.com/f.json"
        )
        client.searchResult = .success(TestFixtures.apiResponse(items: [other]))
        client.fetchImagesResult = .success(["http://images.example.com/medium.jpg"])

        let service = ImageProviderService(client: client)
        let result = try await service.loadTodaysImage()

        XCTAssertEqual(result.item.data?.first?.title, "Fallback")
        XCTAssertEqual(result.imageURLs, ["https://images.example.com/medium.jpg"])
    }

    func testThrowsNoItems() async {
        let client = MockNetworkClient()
        client.searchResult = .success(TestFixtures.apiResponse(items: []))
        let service = ImageProviderService(client: client)

        do {
            _ = try await service.loadTodaysImage()
            XCTFail("Expected noItems")
        } catch let error as Errors.ImageProvider {
            XCTAssertEqual(error, .noItems)
        } catch {
            XCTFail("Unexpected error: \(error)")
        }
    }

    func testThrowsMissingAssetURL() async {
        let client = MockNetworkClient()
        let item = TestFixtures.item(
            dateCreated: TestFixtures.todaysAnniversaryDateCreated,
            href: nil
        )
        client.searchResult = .success(TestFixtures.apiResponse(items: [item]))
        let service = ImageProviderService(client: client)

        do {
            _ = try await service.loadTodaysImage()
            XCTFail("Expected missingAssetURL")
        } catch let error as Errors.ImageProvider {
            XCTAssertEqual(error, .missingAssetURL)
        } catch {
            XCTFail("Unexpected error: \(error)")
        }
    }

    func testRewritesOnlyHTTPScheme() async throws {
        let client = MockNetworkClient()
        let item = TestFixtures.item(
            dateCreated: TestFixtures.todaysAnniversaryDateCreated,
            href: "https://example.com/a.json"
        )
        client.searchResult = .success(TestFixtures.apiResponse(items: [item]))
        client.fetchImagesResult = .success([
            "http://cdn.example.com/large.jpg",
            "https://cdn.example.com/original.jpg"
        ])

        let service = ImageProviderService(client: client)
        let result = try await service.loadTodaysImage()

        XCTAssertEqual(
            result.imageURLs,
            [
                "https://cdn.example.com/large.jpg",
                "https://cdn.example.com/original.jpg"
            ]
        )
    }
}
